### PR TITLE
Add register user

### PR DESCRIPTION
**Pull request by huwshimi**

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6888/

**Comment by huwshimi** [comment]
QA: see me in my office.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6889/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6890/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6892/

File: development.ini [hunk 1]
```diff
@@ -47,21 +47,25 @@ jujugui.interactive_login = true
...
 # Production
 ;jujugui.charmstore_url = https://api.jujucharms.com/charmstore/
 ;jujugui.plans_url = https://api.jujucharms.com/omnibus/
+;jujugui.register_user_url = https://api.jujucharms.com/payment/
 ;jujugui.terms_url = https://api.jujucharms.com/terms/
 
 # jujugui.org
 ;jujugui.charmstore_url = https://api.jujugui.org/charmstore/
 ;jujugui.plans_url = https://api.jujugui.org/omnibus/
+;jujugui.register_user_url = https://api.jujugui.org/payment/
```
Comment (frankban) [inline development.ini]: Those jujugui.org URLs are all wrong. While you're here, please replace api.jujugui.org with www.jujugui.org everywhere.

File: jujugui/static/gui/src/app/jujulib/register-user.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+      @public getUser
+      @param name {String} The user's username.
+      @param callback {Function} A callback to handle errors or accept the
+        data from the request. Must accept an error message or null as its
+        first parameter and the user data as its second. The user data
+        includes the following fields:
+          - nickname {String} The user's nickname
+          - first {String} The user's first name
+	        - last {String} The user's first name
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/register-user.js]: Indentation weirdness?

Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/register-user.js]: Also s/first/last/ here

File: jujugui/static/gui/src/app/jujulib/register-user.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	        - email {String} The user's email address
+	        - business {Boolean} whether this is a business account
+          - addresses {Array} A list of address objects (if defined)
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/register-user.js]: What's an address object? What fields can we expect here?

File: jujugui/static/gui/src/app/jujulib/register-user.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+          - vat {String} The VAT number (if defined)
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/register-user.js]: or null?

File: jujugui/static/gui/src/app/jujulib/register-user.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+          - business-name {String} The business name (if defined)
+	        - billing-addresses {Array} A list of billing address objects
+              (if defined)
+	        - payment-methods {Array} A list of payment method objects
+          - allow-email {Boolean} Whether the user allows emails
+          - valid {Boolean} Whether the user is valid
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/register-user.js]: When a user is not valid? Please explain.

File: jujugui/static/gui/src/app/jujulib/register-user.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        If the user is not found, the second argument is null.
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/register-user.js]: Really nice docstrings in this module, thanks!

File: jujugui/static/gui/src/app/jujulib/register-user.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    },
+
+    /**
+      Add a user.
+
+      @public createUser
+      @param name {String} The user's username.
+      @param user {Object} The user data object, containing:
+        - first {String} The user's first name
+        - last {String} The user's first name
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/register-user.js]: s/first/last/

File: jujugui/static/gui/src/app/jujulib/register-user.js [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        - allowEmail {Boolean} Whether the user allows emails
+        - token {String} A Stripe token (not required)
+        - paymentMethodName {String} The name of the payment method
+            (not required)
+      @param callback {Function} A callback to handle errors or accept the
+        data from the request. Must accept an error message or null as its
+        first parameter and a user object as its second.
+    */
+    createUser: function(name, user, callback) {
+      const handler = function(error, user) {
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/register-user.js]: It seems that we have the same handler for both API calls? If so, could this handler be abstracted out?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6899/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6902/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6904/

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**